src/envoy/tcp/sni_verifier/sni_verifier.cc

```diff
@@ -0,0 +1,161 @@
+/* Copyright 2018 Istio Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "src/envoy/tcp/sni_verifier/sni_verifier.h"
+
+#include "envoy/buffer/buffer.h"
+#include "envoy/common/exception.h"
+#include "envoy/network/connection.h"
+#include "envoy/stats/scope.h"
+
+#include "common/common/assert.h"
+
+#include "openssl/bytestring.h"
+#include "openssl/ssl.h"
+
+namespace Envoy {
+namespace Tcp {
+namespace SniVerifier {
+
+Config::Config(Stats::Scope& scope, uint32_t max_client_hello_size)
+    : stats_{SNI_VERIFIER_STATS(POOL_COUNTER_PREFIX(scope, "sni_verifier."))},
+      ssl_ctx_(SSL_CTX_new(TLS_with_buffers_method())),
+      max_client_hello_size_(max_client_hello_size) {
+  if (max_client_hello_size_ > TLS_MAX_CLIENT_HELLO) {
+    throw EnvoyException(fmt::format(
+        "max_client_hello_size of {} is greater than maximum of {}.",
+        max_client_hello_size_, size_t(TLS_MAX_CLIENT_HELLO)));
+  }
+
+  SSL_CTX_set_options(ssl_ctx_.get(), SSL_OP_NO_TICKET);
+  SSL_CTX_set_session_cache_mode(ssl_ctx_.get(), SSL_SESS_CACHE_OFF);
+  SSL_CTX_set_tlsext_servername_callback(
+      ssl_ctx_.get(), [](SSL* ssl, int* out_alert, void*) -> int {
+        SniVerifierFilter* filter =
+            static_cast<SniVerifierFilter*>(SSL_get_app_data(ssl));
+        filter->onServername(
+            SSL_get_servername(ssl, TLSEXT_NAMETYPE_host_name));
+
+        // Return an error to stop the handshake; we have what we wanted
+        // already.
+        *out_alert = SSL_AD_USER_CANCELLED;
+        return SSL_TLSEXT_ERR_ALERT_FATAL;
+      });
+}
+
+bssl::UniquePtr<SSL> Config::newSsl() {
+  return bssl::UniquePtr<SSL>{SSL_new(ssl_ctx_.get())};
+}
+
+thread_local uint8_t SniVerifierFilter::buf_[Config::TLS_MAX_CLIENT_HELLO];
```

Comment:
Is it possible that two filter instances from two downstream connections are sharing this buffer so their data collapse?

Reply:
@qiwzhang My understanding is that `thread_local` prevents it. This filter is modeled after Envoy's tls inspector, it also uses `thread_local` - https://github.com/envoyproxy/envoy/blob/master/source/extensions/filters/listener/tls_inspector/tls_inspector.cc#L63.

An explanation about `thread_local` on StackOverflow https://stackoverflow.com/a/11984017/553720

Reply:
Correct; one network filter instance per downstream connection, if you do `thread_local` you are will be sharing this buffer between them.

Reply:
@htuch and for a listener filter, like tls inspector, no such sharing will occur, will it?

Reply:
@vadimeisenbergibm sharing will happen as long as the new connection being processesd in the listener filter is on the same worker thread.

Reply:
@htuch I see, now it is clear, thanks!